### PR TITLE
Improve Hive JSON writer performance

### PR DESCRIPTION
**Pull request by pettyjamesm**
## Description
Avoids dynamically switching over each field and nested field type for each row of input when writing JSON output files by constructing type specific field writers for each output field and using virtual dispatch instead.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Hive
* Improve performance of native Hive JSON file writer 
```

### CODE REVIEW

**Comment by dain** [comment]
Can you describe logically what the change is... it is hard to follow the code diff

**Comment by pettyjamesm** [comment]
> Can you describe logically what the change is... it is hard to follow the code diff

Logically, we’re avoiding a per-row traversal of the types (and nested complex types) to determine which generator methods to call. 

Previously we were traversing the big if/else per field type comparison chain once per row, which could reach inlining limits for complex types easily via recursion.

With this change we’re doing that traversal once at the time of writer creation to build up instances of value writers that just use virtual dispatch for each row which benefits writers with lots of columns (and especially complex types that otherwise would recurse).

**Comment by pettyjamesm** [comment]
@dain - I made some minor code organization tweaks in the latest revision to apply the same strategy for map keys and moved some code around to make the code more diff-friendly when you're reviewing. Should be ready to review now.

**Review by dain — APPROVED** [review]
Looks good!
